core/cache-handler.mjs

```diff
@@ -1,11 +1,236 @@
+			return;
+		}
+		await client.unlink(keysToDelete);
+		const updateTtlOperation = client.hdel(keyPrefix + sharedTagsTtlKey, tagsAndTtlToDelete);
+		const updateTagsOperation = client.hdel(keyPrefix + sharedTagsKey, tagsAndTtlToDelete);
+		await Promise.all([updateTagsOperation, updateTtlOperation]);
+	}
+	const revalidatedTagsKey = keyPrefix + REVALIDATED_TAGS_KEY;
+	return {
+		name: "pubpub-redis-strings",
+		async get(key, { implicitTags }) {
+			const result = await client.get(keyPrefix + key);
+			if (!result) {
+				return null;
+			}
+			const cacheValue = JSON.parse(result);
+			if (!cacheValue) {
+				return null;
+			}
+			const sharedTagKeyExists = await client.hexists(keyPrefix + sharedTagsKey, key);
+			if (!sharedTagKeyExists) {
+				await client.unlink(keyPrefix + key);
+				return null;
+			}
+			const combinedTags = /* @__PURE__ */ new Set([...cacheValue.tags, ...implicitTags]);
+			if (combinedTags.size === 0) {
+				return cacheValue;
+			}
+			const revalidationTimes = await client.hmget(
+				revalidatedTagsKey,
+				Array.from(combinedTags)
+			);
+			for (const timeString of revalidationTimes) {
+				if (timeString && Number.parseInt(timeString, 10) > cacheValue.lastModified) {
+					await client.unlink(keyPrefix + key);
+					return null;
+				}
+			}
+			return cacheValue;
+		},
+		async set(key, cacheHandlerValue) {
+			const lifespan = cacheHandlerValue.lifespan;
+			const setTagsOperation =
+				cacheHandlerValue.tags.length > 0
+					? client.hset(
+							keyPrefix + sharedTagsKey,
+							key,
+							JSON.stringify(cacheHandlerValue.tags)
+						)
+					: void 0;
+			const setSharedTtlOperation = lifespan
+				? client.hset(keyPrefix + sharedTagsTtlKey, key, lifespan.expireAt)
+				: void 0;
+			await Promise.all([setTagsOperation, setSharedTtlOperation]);
+			if (typeof lifespan?.expireAt === "number") {
+				await client.set(
+					keyPrefix + key,
+					JSON.stringify(cacheHandlerValue),
+					"EXAT",
+					lifespan.expireAt
+				);
+			} else {
+				await client.set(keyPrefix + key, JSON.stringify(cacheHandlerValue));
+			}
+		},
+		async revalidateTag(tag) {
+			if (isImplicitTag(tag)) {
+				await client.hset(revalidatedTagsKey, tag, Date.now());
+			}
+			await Promise.all([revalidateTags(tag), revalidateSharedKeys()]);
+		},
+		async delete(key) {
+			await client.unlink(keyPrefix + key);
+			await Promise.all([
+				client.hdel(keyPrefix + sharedTagsKey, key),
+				client.hdel(keyPrefix + sharedTagsTtlKey, key),
+			]);
+		},
+	};
+}
+
+async function getCacheHandlerPromise() {
+	let redisClient = null;
+	if (PHASE_PRODUCTION_BUILD !== process.env.NEXT_PHASE) {
+		try {
+			redisClient = new Redis({
+				host: process.env.VALKEY_HOST,
+				lazyConnect: true,
+				commandTimeout: 1000,
+				retryStrategy: (times) => {
+					console.log("Retrying redis connection attempt:", times);
+					return (2 ^ times) + Math.random() * 1000;
+				},
+			});
+
+			await redisClient.connect();
```

Comment:
I'd love to reuse the singleton client from `lib/redis` but I can't seem to import it in this file (lib/redis.ts gets inlined/chunked somewhere, and the build process probably doesn't look at imports in this file). So we have to repeat some of the connection logic + params, and we have two clients per instance, but at least it's not more!